### ISSUE TITLE
Multiple sources of the same type to one destination - aka $.extend

### ISSUE THREAD

**Issue by andrewm1986** [AutoMapper/AutoMapper#1015]
### TL;DR

jQuery.extend (https://api.jquery.com/jquery.extend/) like functionality would be a really good addition to AutoMapper.

### Detail
I quite often have a need to merge multiple objects of the same type back into either the same type or a different type.

I have an example now where I need to merge multiple objects of type `Insurance` into my `InsuranceViewModel`

```c#
public class Insurance 
{
    public decimal? Premium { get; set; }
    public decimal? OutOfPocket { get; set; }
}

public class InsuranceViewModel
{
    public decimal? Premium { get; set; }
    public decimal? OutOfPocket { get; set; }
}
```

Behind the scenes a user's insurance level is an aggregation of different insurances that may or may not have various properties set.

Currently with AutoMapper I would do something like this

```c#
var insurance1 = new Insurance { Premium = 123 };
var insurance2 = new Insurance { OutOfPocket = 456 };

InsuranceViewModel viewModel = Mapper.Map<InsuranceViewModel>(insurance1);
Mapper.Map<Insurance, InsuranceViewModel>(insurance2);
```

However, in my example that would result in `viewModel` having a `Premium` of `null`, because it overwrites the `Premium` value from `insurance1`.

I suggest a `Mapper.Merge` method that works exactly the same as `Mapper.Map` but with no differences

1. Overloads for accepting an array of source.
2. When the value it is going to overwrite the destination property with is `null` it does not overwrite the value.

This would allow me to write the above as

```c#
var insurance1 = new Insurance { Premium = 123 };
var insurance2 = new Insurance { OutOfPocket = 456 };

InsuranceViewModel viewModel = Mapper.Merge<InsuranceViewModel>(new[] { insurance1, insurance2 });

// or

InsuranceViewModel viewModel = Mapper.Merge<InsuranceViewModel>(insurance1, insurance2);


```

`viewModel` would now have `Premium` set to `123` and `OutOfPocket` set to `456`

If this is something that the project wants then I would be happy to look at doing the work myself.  I would quite enjoy it.

**Comment by jbogard** [comment]
Give it a go! I like using a different named mapping method.

**Comment by andrewm1986** [comment]
Rightio.  I'll open a PR

**Comment by lbargaoanu** [comment]
@andrewm1986 Do you still want to do this?

**Comment by andrewm1986** [comment]
Sorry, I've been out of he country for the past few weeks!

I'm due back a week today.  I'll pick this up then and we can talk through my approach on the PR I'll create

**Comment by jbogard** [comment]
Closing for now...
